### PR TITLE
Vip4: add $Restrict in vcc

### PR DESCRIPTION
**Pull request by walid-git**
This PR introduces the $Restrict feature described in [vip4](https://github.com/varnishcache/varnish-cache/wiki/VIP4%3A-Restrict-VMOD-function-call-sites).

It provides the ability to restrict vmod functions and methods scope so that they can only be called
from limited VCL built-in sub routines. The first commit introduces the feature, and the following ones use it on relevant vmods.

### CODE REVIEW

File: doc/sphinx/reference/vmod.rst [hunk 1]
```diff
@@ -264,6 +264,19 @@ managing instances, in particular their memory management. As the
...
+---------------------------------------
+
+The ``$Restrict`` stanza offers a way to limit the scope of the preceding vmod function
+or method, so that they can only be called from restricted vcl call sites.
+It must only appear after a ``$Method`` or ``$Function`` and has the following syntax::
+
+    $Restrict scope1 [scope2 ...]
+
+Possible scope values are:
+backend, client, housekeeping, recv, pipe, pass, hash, purge, miss, hit,
```
Comment (nigoroll) [inline doc/sphinx/reference/vmod.rst:277]: why not name the scopes `vcl_*` except for the "meta scopes" `backend, client, housekeeping`

**Review by nigoroll — APPROVED** [review]
👍🏽 this looks solid to me, thank you @walid-git 
Besides the minor specific comment, I'd suggest to change the error handling in vmod functions/methods which now have a `$Restrict`ed scope into assertions, like for this case:

```diff
diff --git a/vmod/vmod_unix.c b/vmod/vmod_unix.c
index 1edfca9d5..39a7b3307 100644
--- a/vmod/vmod_unix.c
+++ b/vmod/vmod_unix.c
@@ -49,9 +49,6 @@
 #define VERR(ctx, fmt, ...) \
        VSLb((ctx)->vsl, SLT_VCL_Error, "vmod unix error: " fmt, __VA_ARGS__)
 
-#define FAILNOINIT(ctx) \
-       FAIL((ctx), "may not be called in vcl_init or vcl_fini")
-
 #define ERRNOTUDS(ctx) \
        ERR((ctx), "not listening on a Unix domain socket")
 
@@ -93,10 +90,7 @@ vmod_##func(VRT_CTX)                                 \
        int ret;                                        \
                                                        \
        CHECK_OBJ_NOTNULL(ctx, VRT_CTX_MAGIC);          \
-       if ((ctx->method & VCL_MET_TASK_H) != 0) {      \
-               FAILNOINIT(ctx);                        \
-               return (-1);                            \
-       }                                               \
+       AZ(ctx->method & VCL_MET_TASK_H);               \
                                                        \
        sp = get_sp(ctx);                               \
        if (!sp->listen_sock->uds) {                    \
```

File: doc/sphinx/reference/vmod.rst (hunk 1 above)
Comment (walid-git) [reply doc/sphinx/reference/vmod.rst:277]: Thanks for the review. I added the other suggested change. Regarding the naming, I don't have an opinion, we should probably discuss it with the others during bugwash.

Comment (walid-git) [reply doc/sphinx/reference/vmod.rst:277]: Changed the names of scopes as agreed during bugwash to vcl_* (except for backend client and housekeeping)

**Review by nigoroll — APPROVED** [review]

**Pull request merged**